Method is inaccessible from method in nested classes with inheritance.

## Steps to Reproduce

1. Compile Program.cs `mcs Program.cs`
2. Run Program.exe `mono Program.exe`

```c#
namespace ConsoleApplication1
{
    public abstract class SuperClass
    {
        protected abstract class SuperInnerAbstractClass
        {
            protected class SuperInnerInnerClass
            {
            }
        }
    }

    public class ChildClass : SuperClass
    {
        private class ChildInnerClass : SuperInnerAbstractClass
        {
            private readonly SuperInnerInnerClass s_class = new SuperInnerInnerClass();
        }

        public ChildClass()
        {
            var childInnerClass = new ChildInnerClass();
        }
    }
    
    internal class Program
    {
        public static void Main(string[] args)
        {
            new ChildClass();
        }
    }
}
```

<!--
You may drag & drop the attachment (repro code/solution, screenshot, etc.) onto the issue.
-->

## Current Behavior
The code compiles successfully, but when run on mono `System.MethodAccessException` is thrown.

## Expected Behavior
Code should execute without exception.  I have tested this on Windows 10, Visual Studio 2017 and this code executes, and runs without error.

### On which platforms did you notice this

- [x] macOS
- [ ] Linux
- [ ] Windows

**Version Used**:
```
Mono JIT compiler version 5.10.0.176 (2017-12/a7adb18ce2f Wed Mar 14 06:05:24 EDT 2018)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          yes(3.6.0svn-mono-master/8b1520c8aae)
	GC:            sgen (concurrent by default)
```

### Stacktrace

```
Unhandled Exception:
System.MethodAccessException: Method `ConsoleApplication1.SuperClass+SuperInnerAbstractClass+SuperInnerInnerClass..ctor()' is inaccessible from method `ConsoleApplication1.ChildClass+ChildInnerClass..ctor()'
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_throw_method_access(intptr,intptr)
  at ConsoleApplication1.Program.Main (System.String[] args) [0x00001] in /Users/jeff/Projects/ConsoleApplication1/ConsoleApplication1/Program.cs:30 
[ERROR] FATAL UNHANDLED EXCEPTION: System.MethodAccessException: Method `ConsoleApplication1.SuperClass+SuperInnerAbstractClass+SuperInnerInnerClass..ctor()' is inaccessible from method `ConsoleApplication1.ChildClass+ChildInnerClass..ctor()'
  at (wrapper managed-to-native) System.Object.__icall_wrapper_mono_throw_method_access(intptr,intptr)
  at ConsoleApplication1.Program.Main (System.String[] args) [0x00001] in /Users/jeff/Projects/ConsoleApplication1/ConsoleApplication1/Program.cs:30 

```

--- a/mono/metadata/class.c
+++ b/mono/metadata/class.c
@@ -5236,14 +5236,11 @@ can_access_type (MonoClass *access_klass, MonoClass *member_klass)
 	if (is_nesting_type (access_klass, member_klass) || (m_class_get_nested_in (access_klass) && is_nesting_type (m_class_get_nested_in (access_klass), member_klass)))
 		return TRUE;
 
-	MonoClass *member_klass_nested_in = m_class_get_nested_in (member_klass);
-	if (member_klass_nested_in && !can_access_type (access_klass, member_klass_nested_in))
-		return FALSE;
-
 	/*Non nested type with nested visibility. We just fail it.*/
 	if (access_level >= TYPE_ATTRIBUTE_NESTED_PRIVATE && access_level <= TYPE_ATTRIBUTE_NESTED_FAM_OR_ASSEM && m_class_get_nested_in (member_klass) == NULL)
 		return FALSE;
 
+	MonoClass *member_klass_nested_in = m_class_get_nested_in (member_klass);
 	switch (access_level) {
 	case TYPE_ATTRIBUTE_NOT_PUBLIC:
 		return can_access_internals (access_klass_assembly, member_klass_assembly);
@@ -5252,16 +5249,16 @@ can_access_type (MonoClass *access_klass, MonoClass *member_klass)
 		return TRUE;
 
 	case TYPE_ATTRIBUTE_NESTED_PUBLIC:
-		return TRUE;
+		return member_klass_nested_in && can_access_type (access_klass, member_klass_nested_in);
 
 	case TYPE_ATTRIBUTE_NESTED_PRIVATE:
-		return is_nesting_type (member_klass, access_klass);
+		return is_nesting_type (member_klass, access_klass) && member_klass_nested_in && can_access_type (access_klass, member_klass_nested_in);
 
 	case TYPE_ATTRIBUTE_NESTED_FAMILY:
 		return mono_class_has_parent_and_ignore_generics (access_klass, m_class_get_nested_in (member_klass)); 
 
 	case TYPE_ATTRIBUTE_NESTED_ASSEMBLY:
-		return can_access_internals (access_klass_assembly, member_klass_assembly);
+		return can_access_internals (access_klass_assembly, member_klass_assembly) && member_klass_nested_in && can_access_type (access_klass, member_klass_nested_in);
 
 	case TYPE_ATTRIBUTE_NESTED_FAM_AND_ASSEM:
 		return can_access_internals (access_klass_assembly, m_class_get_image (member_klass_nested_in)->assembly) &&

--- a/mono/tests/Makefile.am
+++ b/mono/tests/Makefile.am
@@ -539,7 +539,8 @@ TESTS_CS_SRC=		\
 	tailcall-generic-cast-cs.cs \
 	tailcall-interface.cs \
 	sizeof-empty-structs.cs \
-	bug-60843.cs
+	bug-60843.cs	\
+	nested_type_visibility.cs
 
 if AMD64
 TESTS_CS_SRC += async-exc-compilation.cs finally_guard.cs finally_block_ending_in_dead_bb.cs

--- a/mono/tests/nested_type_visibility.cs
+++ b/mono/tests/nested_type_visibility.cs
@@ -1,0 +1,36 @@
+namespace ConsoleApplication1
+{
+    public abstract class SuperClass
+    {
+        protected abstract class SuperInnerAbstractClass
+        {
+            protected class SuperInnerInnerClass
+            {
+            }
+        }
+    }
+
+    public class ChildClass : SuperClass
+    {
+        private class ChildInnerClass : SuperInnerAbstractClass
+        {
+            private readonly SuperInnerInnerClass s_class = new SuperInnerInnerClass();
+        }
+
+        public ChildClass()
+        {
+            var childInnerClass = new ChildInnerClass();
+        }
+    }
+    
+    internal class Program
+    {
+        public static int Main(string[] args)
+        {
+            new ChildClass();
+
+			return 0;
+        }
+    }
+}
+

